doc/developer/design/20231027_refresh_mvs.md

```diff
@@ -0,0 +1,279 @@
+
+### Implementation Alternative: Creating a New Dataflow at Each Refresh
+
+(The above alternatives are alternatives to `REFRESH EVERY` MVs, but this alternative is just about the implementation details of `REFRESH EVERY` MVs.)
+
+This approach would need similar replica management as the proposed solution, but instead of rehydrating an existing dataflow, we would always create a new dataflow to run the refresh. An advantage of this approach is that these dataflows could be single-time dataflows, and thus rely on one-shot SELECT optimizations. However, this approach would be more complicated, as we would have to separate the lifecycle of dataflows from the lifecycle of MVs. These new dataflows would be a kind of hybrid between Peek dataflows and MV dataflows, and would require new code in various parts of the Compute Controller. As we are under some time pressure to implement this feature, we opted for the simpler solution proposed in this document.
+
+### Implementation Alternative: Suspending the Replica Process to Disk
+
+As mentioned in the [scoping section](#out-of-scope), an alternative implementation that would cover also intermediate refresh intervals would be to suspend the replica process and save its entire state to S3 and a local disk cache, and restore it when a refresh is needed. [There are existing techniques for saving process state, e.g., CRIU.](https://faun.pub/kubernetes-checkpointing-a-definitive-guide-33dd1a0310f6) However, this would be a much bigger project than the proposed solution.
+
+## Automated Cluster Scheduling for `REFRESH EVERY` MVs
+
+TODO: updated discussion
+
+## Introspection / Observability
+
+For showing the `REFRESH` options of each materialized view, I'm planning to create a new table in `mz_internal` called `mz_materialized_view_refresh_options`, which will have a row for each non-trivial refresh option of each MV. The columns will be `(materialized_view_id text, refresh_type text, refresh_interval interval, refresh_interval_aligned_to timestamptz, refresh_at timestamptz)`. In `refresh_type` we would show either the string "REFRESH EVERY" or "REFRESH AT". From those columns that show details of a particular refresh option, only those columns will have a non-null value that are relevant for the refresh option that the row is about: a `REFRESH EVERY` would be non-null in `interval` and `interval_aligned_to`, a `REFRESH AT` would be non-null in `refresh_at`. `REFRESH ON COMMIT` is considered a trivial refresh option, and won't show up in the table. (See Nikhil's [msg](https://materializeinc.slack.com/archives/C06535JL58R/p1701228087745239?thread_ts=1700595291.424189&cid=C06535JL58R) on an old design thread.)
```

Comment:
I don't think there's precedent for a column like `refresh_type` having values that look like SQL SYNTAX like `REFRESH EVERY` or `REFRESH AT`. I think it would be more in line with what do elsewhere to use kebab-case, so `every` or `at` as the type. That also admits an easy way to represent `REFRESH = ON COMMIT`, by using `on-commit` as the refresh type. Now that I'm looking at it again, I'm thinking it's probably quite a bit nicer to represent `on-commit` with an explicit row, rather than it being implied if no refresh options exist for a given view.

Also, not sure about having the `refresh_` prefix in the name of each column. It's plausibly implied by the name of the view (`..._refresh_options`), and in other tables we don't include in the column names information that's included in the relation's name.

Reply:
> I don't think there's precedent for a column like refresh_type having values that look like SQL SYNTAX like REFRESH EVERY or REFRESH AT. I think it would be more in line with what do elsewhere to use kebab-case, so every or at as the type.

Ok!

> Now that I'm looking at it again, I'm thinking it's probably quite a bit nicer to represent on-commit with an explicit row, rather than it being implied if no refresh options exist for a given view.

Agreed!

> Also, not sure about having the refresh_ prefix in the name of each column. It's plausibly implied by the name of the view (..._refresh_options), and in other tables we don't include in the column names information that's included in the relation's name.

Yes, the prefix can be considered redundant. I'll remove it. (Originally, I didn't have those, but added it after [this comment](https://github.com/MaterializeInc/materialize/pull/22776#discussion_r1561115426).) 

> I'm not sold on the name of the view, though — which maybe ties into the bigger question of how we present this feature to users. I think it's more intuitive to refer to each option as a materialized view's "refresh strategy", and this is AFAICT the term Field Engineering has been using. I'd suggest we name this mz_materialized_view_refresh_strategies instead (despite it being a mouthful).

I'm ok with either `mz_materialized_view_refresh_strategies` or `mz_materialized_view_refresh_options`. I'll change it to `mz_materialized_view_refresh_strategies` for now.

Reply:
Oh, when removing the `refresh_` prefix from the column names, I realized that a column name `interval` is slightly inconvenient to use: You have to put it like `[interval]` to avoid the keyword-colname mixup. Is simply having `interval` as the column name still ok?

Reply:
My only immediate concern was exactly that `interval` would be a reserved keyword, which it is according to the SQL standard, but not according to PostgreSQL ([docs](https://www.postgresql.org/docs/current/sql-keywords-appendix.html)). We seem to digress here:

```sql
postgres=# CREATE TABLE test (interval int);
CREATE TABLE
postgres=# SELECT * FROM test;
 interval
----------
(0 rows)

postgres=# SELECT interval FROM test;
 interval
----------
(0 rows)
```

```sql
materialize=> CREATE TABLE test (interval int);
CREATE TABLE
materialize=> SELECT * FROM test;
 interval
----------
(0 rows)

materialize=> SELECT interval FROM test;
ERROR:  Expected literal string, found FROM
LINE 1: SELECT interval FROM test;
```

It's likely that this table won't be directly queried by users (assuming we expose this information in the console), but I think it'd be safer (and clearer) to keep the `refresh` prefix in this case.

Reply:
Yeah, I'm also leaning towards putting back the `refresh_` prefix in this particular case. @benesch, what do you think?

Reply:
I'd prefer to just make our handling of `interval` match PostgreSQL! I put up https://github.com/MaterializeInc/materialize/pull/26721 to do so.